tools/approval-service/internal/service/ttleventcache_test.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	ecNegative := NewTTLEventCache(-5 * time.Second)
+	if ecNegative.ttl != defaultTTL {
+		t.Errorf("NewTTLEventCache(-5) did not set default TTL: got: %v; want: %v", ecNegative.ttl, defaultTTL)
+	}
+}
+
+func TestTTLEventCache_ValidTTL_ShouldOverrideDefault(t *testing.T) {
+	// Arrange
+	oneSecondTTlForTest := 1 * time.Second
+
+	// Act
+	ecPositive := NewTTLEventCache(oneSecondTTlForTest)
+
+	// Assert
+	if ecPositive.ttl != oneSecondTTlForTest {
+		t.Errorf("NewTTLEventCache(1) did not set correct TTL: got: %v; want: %v", ecPositive.ttl, oneSecondTTlForTest)
+	}
+}
+
+func TestTTLEventCache_TryAdd_HappyPath(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		// Arrange
+		keyForTest := "key-for-test"
+		ec := NewTTLEventCache(15 * time.Second)
+
+		// Act
+		isSuccessfulFirstTryAdd := ec.TryAdd(keyForTest)
+
+		// Assert
+		if !isSuccessfulFirstTryAdd {
+			t.Errorf("TryAdd(key) returned false on first add, expected true")
+		}
+		ec.checkInternalState(t, 1, 1)
+	})
+}
+
+func TestTTLEventCache_TryAdd_BasicRejectDuplicates(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ec := NewTTLEventCache(15 * time.Second)
+		keyForTest := "key-for-test"
+
+		isSuccessfulFirstTryAdd := ec.TryAdd(keyForTest)
+		if !isSuccessfulFirstTryAdd {
+			t.Errorf("TryAdd(key) returned false on first add, expected true")
+		}
+		ec.checkInternalState(t, 1, 1)
+
+		isSuccessfulSecondTryAdd := ec.TryAdd(keyForTest)
+		if isSuccessfulSecondTryAdd {
+			t.Errorf("TryAdd(keyForTest) returned true on duplicate add, expected false")
+		}
+		ec.checkInternalState(t, 1, 1)
+	})
+}
+
+func TestTTLEventCache_TryAdd_Concurrency(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ttl := 15 * time.Second
+		ec := NewTTLEventCache(ttl)
+
+		const (
+			distinctKeys   = 5
+			requestsPerKey = 200
+		)
+
+		var successfulAdds int32
+
+		// Schedule one small step per logical request. The synctest harness runs these
+		// steps under a deterministic scheduler so inter-leavings are reproducible.
+		for i := 0; i < distinctKeys; i++ {
+			key := fmt.Sprintf("key-%d", i)
+			for j := 0; j < requestsPerKey; j++ {
+				if ec.TryAdd(key) {
+					atomic.AddInt32(&successfulAdds, 1)
+				}
+			}
+		}
+
+		// After all scheduled steps run (all requests within the TTL), exactly one
+		// successful add per key is expected.
+		got := int(atomic.LoadInt32(&successfulAdds))
+		if got != distinctKeys {
+			t.Fatalf("expected %d successful adds (one per key), got %d", distinctKeys, got)
+		}
+	})
+
+}
+
+func TestTTLEventCache_TryAdd_AddAfterExpiry(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ttl := 15 * time.Second
+		ec := NewTTLEventCache(ttl)
+		keyABCForTest := "ABC"
+		key123ForTest := "123"
+
+		isSuccessfulKeyOneFirstTry := ec.TryAdd(keyABCForTest)
+		if !isSuccessfulKeyOneFirstTry {
+			t.Errorf("TryAdd(keyOne) returned false on first add, expected true")
+		}
+
+		// wait for TTL to pass (virtual sleep)
+		time.Sleep(ttl + 10*time.Millisecond)
```

Comment:
You do not need the +10ms here. You should be able to sleep the exact TTL due to being in a synctest bubble. This makes the testing more precise and you are able to test boundary conditions should they matter.

Same in the other tests.